drivers/net/mlx5/mlx5_vdpa.c

```diff
@@ -147,14 +160,92 @@ mlx5_vdpa_query_virtio_caps(struct vdpa_priv *priv)
 	 * For now only set protocol features support
 	 */
 	priv->caps.max_num_virtqs = MLX5_VDPA_SW_MAX_VIRTQS_SUPPORTED;
-	priv->caps.virtio_net_features = (1ULL << VHOST_USER_F_PROTOCOL_FEATURES);
+	/*
+	 * TODO (shahafs): Take from QUERY HCA CAP Device Emulation Capabilities.
+	 * For now only set protocol features support
+	 */
+	priv->caps.notify_stride = MLX5_VDPA_SW_NOTIFY_STRIDE;
+	priv->caps.virtio_net_features = MLX5_VDPA_FEATURES;
+	priv->caps.virtio_protocol_features = MLX5_VDPA_PROTOCOL_FEATURES;
 	DRV_LOG(DEBUG, "Virtio Caps:");
 	DRV_LOG(DEBUG, "	dump_mkey=0x%x ", priv->caps.dump_mkey);
 	DRV_LOG(DEBUG, "	max_num_virtqs=0x%x ", priv->caps.max_num_virtqs);
 	DRV_LOG(DEBUG, "	features_bits=0x%" PRIx64, priv->caps.virtio_net_features);
 	return 0;
 }
 
+#define MLX5_IB_MMAP_CMD_SHIFT 8
+#define MLX5_IB_MMAP_CMD_MASK ((1 << MLX5_IB_MMAP_CMD_SHIFT) - 1)
+#define MLX5_IB_CMD_SIZE 8
+#define MLX5_IB_MMAP_VIRTIO_NOTIFY 9
+static inline int
+mlx5_vdpa_get_notify_offset(uint16_t offset)
+{
+	uint16_t ext_offset = MLX5_IB_MMAP_CMD_SHIFT + MLX5_IB_CMD_SIZE;
+	return (((offset >> MLX5_IB_MMAP_CMD_SHIFT) << ext_offset) |
+		(MLX5_IB_MMAP_VIRTIO_NOTIFY << MLX5_IB_MMAP_CMD_SHIFT) |
+		(offset & MLX5_IB_MMAP_CMD_MASK));
+}
+
+static int
+mlx5_vdpa_report_notify_area(int vid, int qid, uint64_t *offset,
+			     uint64_t *size)
+{
+	int dev_id;
+	struct vdpa_priv_list *list;
+	long page_size = sysconf(_SC_PAGESIZE);
+	uint16_t notify_stride;
+
+	dev_id = rte_vhost_get_vdpa_device_id(vid);
+	if (dev_id < 0)
+		goto error;
+	list = find_priv_resource_by_did(dev_id);
+	if (!list)
+		goto error;
+	notify_stride = list->priv->caps.notify_stride;
+	*offset = 0 | mlx5_vdpa_get_notify_offset(qid * notify_stride);
```

Comment:
why do we need the "0 |"?
It makes the code less readable IMO

Reply:
just to not rely on the compiler to clear the rest of the offset bits. 

i can remove it.